ARCHITECTURE.md

```diff
@@ -0,0 +1,73 @@
+# Exonum Architecture
+
+This document describes the high-level architecture of Exonum. Its target
+audience is the developers of Exonums core. To learn how to *use* Exonum, see
+https://exonum.com/doc/.
+
+This document is provided on the best effort bases -- it is not guranteed to be
+up-to-date and complete. The purpose of this document is to highlight the most
+important types, modules and functions in the code, to allow new contributors to
+dive into Exonum quickly. It is assumed that the reader is familiar with core
+concepts of Exonum, such as services and transactions.
+
+
+## Code organization
+
+The repository is a single Cargo workspace of several related packages:
+
+- `exonum`, most of the code lives here,
+
+- `sandbox`, where integration tests live,
+
+- `testkit`, a framework for testing Exonum services.
+
+
+This document currently describes only the `exonum` crate.
+
+
+## Storage
+
+Storage module can be found at `src/storage`. It provides an abstracton over an
+embeded key value store: the `Database` trait. The keys and values of are just
+slices of bytes: `&[u8]`.
+
+To transform raw bytes into native Rust datastructures, the `StorageKey` and
+`StorageValue` traits are used.
+
+On top of raw storage, collections like sets, lists and maps are provided: see
```

Comment:
The fact that everyone names these things as they want (collections, tables, views, indices), tells us that an "index" is a terrible name.

I welcome everyone aboard: https://jira.bf.local/browse/ECR-437

Reply:
➕  💯